### ISSUE TITLE
New ptest failure in tst_rec_vars

### ISSUE THREAD

**Issue by ArchangeGabriel** [Parallel-NetCDF/PnetCDF#72]
Short log:
```
*** TESTING C   tst_rec_vars for record variables to NetCDF4 file  ------ Error at line 78 of tst_rec_vars.c: expect dim X len 4 but got 3
Error at line 78 of tst_rec_vars.c: expect dim X len 4 but got 3
fail with 2 mismatches
```

Most important changes is upgrading netcdf to 4.8.0, but I need to retry again 4.7.4 to verify that this is effectively the culprit.

**Comment by wkliao** [comment]
I believe this is caused by a NetCDF 4.8.0 bug. Below is the same program but
rewritten in NetCDF that can reproduce the same error if compiled with 4.8.0.
Note the variable is created and set with collective access mode, assuming
that mode should guarantee the metadata consistency across all MPI processes.
```
% cat tst_nc4.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <netcdf.h>
#include <netcdf_par.h>

#define CHECK_ERR { \
    if (err != NC_NOERR) { \
        printf("Error at line=%d: %s\n", __LINE__, nc_strerror(err)); \
        return 0; \
    } \
}

int main(int argc, char** argv) {
    int err=NC_NOERR, rank, nprocs;
    int ncid, cmode, varid, dimid, buf;
    size_t start[1], count[1], nrecs;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    buf = rank;

    cmode = NC_CLOBBER | NC_NETCDF4;
    err = nc_create_par("testfile.nc", cmode, MPI_COMM_WORLD, MPI_INFO_NULL,
                        &ncid); CHECK_ERR

    err = nc_def_dim(ncid, "time", NC_UNLIMITED, &dimid); CHECK_ERR
    err = nc_def_var(ncid, "var", NC_INT, 1, &dimid, &varid); CHECK_ERR
    err = nc_var_par_access(ncid, varid, NC_COLLECTIVE); CHECK_ERR
    err = nc_enddef(ncid); CHECK_ERR

    start[0] = rank;
    count[0] = 1;
    err = nc_put_vara_int(ncid, varid, start, count, &buf); CHECK_ERR
    err = nc_inq_dimlen(ncid, dimid, &nrecs); CHECK_ERR

    if (nrecs != nprocs) {
        printf("Rank %d error at line %d of file %s:\n",rank,__LINE__,__FILE__);
        printf("\tafter writing start=%zd count=%zd\n", start[0], count[0]);
        printf("\texpecting number of records = %lld but got %lld\n",
               nprocs, nrecs);
    err = nc_close(ncid); CHECK_ERR

err_out:
    MPI_Finalize();
    return (err < 0);
}
```
Compile and run commands I used.
```
% mpicc -g -O0 tst_nc4.c -o tst_nc4 \
      -I $HOME/NetCDF/4.8.0/include \
      -L $HOME/NetCDF/4.8.0/lib -lnetcdf

% mpiexec -n 4 ./tst_nc4
Rank 0 error at line 40 of file tst_nc4.c:
	after writing start=0 count=1
	expecting number of records = 4 but got 1
Rank 1 error at line 40 of file tst_nc4.c:
	after writing start=1 count=1
	expecting number of records = 4 but got 2
Rank 2 error at line 40 of file tst_nc4.c:
	after writing start=2 count=1
	expecting number of records = 4 but got 3
```

**Comment by ArchangeGabriel** [comment]
Opened https://github.com/Unidata/netcdf-c/issues/2038 to get the netCDF devs opinion on this. ;)

**Comment by wkliao** [comment]
I suggest to not enable netcdf4 feature when building PnetCDF for ArchLinux release.

**Comment by ArchangeGabriel** [comment]
As GitHub auto-linked, that’s what I’ve done for now. ;)

**Comment by ArchangeGabriel** [comment]
Fixed in NetCDF 4.9. :)
